pkg/http/handler/timeout.go

```diff
@@ -75,8 +75,11 @@ func (h *timeToFirstByteTimeoutHandler) ServeHTTP(w http.ResponseWriter, r *http
 		h.handler.ServeHTTP(tw, r.WithContext(ctx))
 	}()
 
-	timeout := time.NewTimer(h.timeout)
-	defer timeout.Stop()
+	timeout := getTimer(h.timeout)
+	var timeoutExpired bool
```

Comment:
fwiw I think another approach here would be to do `if !timeout.Stop() { <- timeout.C }` on line 86, then the channel is always consumed by the time of the defer so we wouldn't need this variable or the extra parameter/logic in `returnTimer` (and the defer can be `defer returnTimer(timeout)` since there's no problem about parameter evaluation). Don't know if that's all that much better, but checking you've considered that approach too.

Reply:
I'm not sure there's no race in that. The docs specifically call out

> To ensure the channel is empty after a call to Stop, check the return value and drain the channel. For example, assuming the program has not received from t.C already:

It's that "assuming the program has not received from t.C already" that trips me up and makes me want to actually enforce that invariant and handle it explicitly. So we'd have to check the bool in there as well. I personally think the current method is more encapsulated and handled "easier" through the `defer`, though I agree that the extra closure is a little ugly.

Reply:
Im trying to convince myself there's not still a race here between the two select cases. Stop only returns true if the call stops the timer, so it doesn't return true if the timer was already expired. But that means -- I think -- the timer might be expired, and may have placed a message on t.C, but the select may have gone for the done arm instead (since it's non-deterministic). In that case don't we end up with a non-drained channel in the timer?

Is Stop smart enough to know whether we've read from t.C yet? Seems hard to see how it could since C is a [buffered channel](https://cs.opensource.google/go/go/+/refs/tags/go1.16.6:src/time/sleep.go;l=87;drc=refs%2Ftags%2Fgo1.16.6;bpv=1;bpt=1) 🤔 

Probably missing something obvious here, but I can't quite convince myself why this works yet

Reply:
Good we're having this discussion. I've been hesitant of sending this PR precisely because of this :joy:.

And you're correct! To make matters worse, https://groups.google.com/g/golang-nuts/c/-8O3AknKpwk suggests there's not a good way to actually make the behavior correct other than checking the last call to `Reset` and the time sent via t.C next to then determine if the interval makes any sense in the context of the last Reset. But... meh. Getting into the realm of "too ugly to make sense" if you ask me.

Reply:
err, actually, doesn't the code currently already account for that? We actually force-read from the channel below if we have **not** gone into the timer's leg.

Funnily https://groups.google.com/g/golang-dev/c/c9UUfASVPoU/m/tlbK2BpFEwAJ suggests just the same solution I opted for (pinkyswear I didn't read this before I implemented it).

Reply:
> Is Stop smart enough to know whether we've read from t.C yet?

Would that be something worth writing a test for?

Reply:
Good point, there is one interesting case that might be worth adding a test for indeed.

Reply:
Actually: It's complicated :joy:. What we're discussing here are races that we can't reliably trigger in a test I think. Only if you like run the test a gazillion times (and/or potentially if you fake the timer).